### PR TITLE
Add parallel randomized benchmarking

### PR DESCRIPTION
**Pull request by eliottrosenberg**

### CODE REVIEW

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 1]
```diff
@@ -230,6 +230,70 @@ def single_qubit_randomized_benchmarking(
...
+        num_circuits: The number of random circuits generated for each
+            number of Cliffords.
+        repetitions: The number of repetitions of each circuit.
+
+    Returns:
+        A dictionary from qubits to RandomizedBenchMarkResult objects.
+    """
+
+    if qubits is None:  # pragma: no cover
+        try:
```
Comment (NoureldinYosri) [inline cirq-core/cirq/experiments/qubit_characterizations.py:263]: I don't understand this block. is this a common pattern we have? could we seprate it into a method (e.g. `qubits_from_sampler`)?

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 2]
```diff
@@ -230,6 +230,70 @@ def single_qubit_randomized_benchmarking(
...
+
+    cliffords = _single_qubit_cliffords()
+    c1 = cliffords.c1_in_xy if use_xy_basis else cliffords.c1_in_xz
+    cfd_mats = np.array([_gate_seq_to_mats(gates) for gates in c1])
+
+    # create circuits
+    circuits_all = []
+    for num_cfds in num_clifford_range:
+        for _ in range(num_circuits):
+            circuits_all.append(_create_parallel_rb_circuit(qubits, num_cfds, c1, cfd_mats))
```
Comment (NoureldinYosri) [inline cirq-core/cirq/experiments/qubit_characterizations.py:277]: to be sure, the inner loop is meant to add a number of slightly different circuits with the same `num_cfds`, right?

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 1 above)
Comment (dstrain115) [reply cirq-core/cirq/experiments/qubit_characterizations.py:263]: Also, let's make sure it is covered by tests.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 3]
```diff
@@ -230,6 +230,70 @@ def single_qubit_randomized_benchmarking(
...
+            number of Cliffords.
+        repetitions: The number of repetitions of each circuit.
+
+    Returns:
+        A dictionary from qubits to RandomizedBenchMarkResult objects.
+    """
+
+    if qubits is None:  # pragma: no cover
+        try:
+            device = sampler.processor.get_device()  # type: ignore
```
Comment (dstrain115) [inline cirq-core/cirq/experiments/qubit_characterizations.py:264]: What's going on with the "type: ignore" commands?  You can say "device: TYPE = ..." in the first declaration to denote what type it should be, if the two device types disagree.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 4]
```diff
@@ -230,6 +230,70 @@ def single_qubit_randomized_benchmarking(
...
+        except:
+            device = sampler.device  # type: ignore
+        qubits = device.metadata.qubit_set
+
+    cliffords = _single_qubit_cliffords()
+    c1 = cliffords.c1_in_xy if use_xy_basis else cliffords.c1_in_xz
+    cfd_mats = np.array([_gate_seq_to_mats(gates) for gates in c1])
```
Comment (dstrain115) [inline cirq-core/cirq/experiments/qubit_characterizations.py:271]: I'd probably avoid abbreviations when possible.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 5]
```diff
@@ -230,6 +230,70 @@ def single_qubit_randomized_benchmarking(
     return RandomizedBenchMarkResult(num_clifford_range, gnd_probs)
 
 
+def parallel_single_qubit_randomized_benchmarking(
+    sampler: 'cirq.Sampler',
+    use_xy_basis: bool = True,
```
[the remark this answers is not in the document]
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:235]: I was following the same format as in `single_qubit_randomized_benchmarking()` in the same file, but I can change it.

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 4 above)
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:271]: I'm using the same variable names as in `single_qubit_randomized_benchmarking()` in the same file.

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 2 above)
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:277]: That's right, different random instances with the same `num_cfds` (number of Clifford gates). It's the same logic as in the `single_qubit_randomized_benchmarking()` function in that file.

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 1 above)
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:263]: Ok, I'll remove this block of code. To get it properly typed and covered by tests would require importing from `cirq_google`. The user will have to input the qubits manually.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 6]
```diff
@@ -230,6 +230,70 @@ def single_qubit_randomized_benchmarking(
...
+    *,
+    qubits: Optional[Iterator['cirq.GridQubit']] = None,
+    num_clifford_range: Sequence[int] = [5, 18, 70, 265, 1000],
```
[the remark this answers is not in the document]
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:238]: This is `np.logspace(np.log10(5), 3, 5, dtype=int)`. I will change it to that.

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 3 above)
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:264]: I'll remove this code block since I would have to import `cirq_google` to get it typed properly.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 7]
```diff
@@ -464,6 +528,18 @@ def _measurement(two_qubit_circuit: circuits.Circuit) -> np.ndarray:
...
 
+def _create_parallel_rb_circuit(
+    qubits: Iterator['cirq.GridQubit'], num_cfds: int, c1: list, cfd_mats: np.ndarray
+) -> 'cirq.Circuit':
+    circuits_to_zip = [_random_single_q_clifford(qubit, num_cfds, c1, cfd_mats) for qubit in qubits]
+    circuit = circuits.Circuit.zip(*circuits_to_zip)
+    measure_moment = circuits.Moment(
+        ops.measure_each(*qubits, key_func=lambda q: f"q{q.row}_{q.col}")  # type: ignore
+    )
+    circuit_with_meas = circuits.Circuit.from_moments(*(list(circuit.moments) + [measure_moment]))
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:532]: Can use more than one argument instead of creating a list first:
```suggestion
    circuit_with_meas = circuits.Circuit.from_moments(*circuit.moments, measure_moment))
```

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 8]
```diff
@@ -464,6 +528,18 @@ def _measurement(two_qubit_circuit: circuits.Circuit) -> np.ndarray:
     return TomographyResult(rho)
 
 
+def _create_parallel_rb_circuit(
+    qubits: Iterator['cirq.GridQubit'], num_cfds: int, c1: list, cfd_mats: np.ndarray
+) -> 'cirq.Circuit':
+    circuits_to_zip = [_random_single_q_clifford(qubit, num_cfds, c1, cfd_mats) for qubit in qubits]
+    circuit = circuits.Circuit.zip(*circuits_to_zip)
+    measure_moment = circuits.Moment(
+        ops.measure_each(*qubits, key_func=lambda q: f"q{q.row}_{q.col}")  # type: ignore
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:530]: If you don't provide a key func, the default will be to use `str(q)` as the key, which should be fine.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 9]
```diff
@@ -230,6 +230,70 @@ def single_qubit_randomized_benchmarking(
...
+    results_all = sampler.run_batch(circuits_all, repetitions=repetitions)
+    gnd_probs: dict = {q: [] for q in qubits}
+    idx = 0
+    for num_cfds in num_clifford_range:
+        excited_probs_l: dict = {q: [] for q in qubits}
+        for _ in range(num_circuits):
+            results = results_all[idx][0]
+            for qubit in qubits:
+                excited_probs_l[qubit].append(
+                    np.mean(results.measurements[f"q{qubit.row}_{qubit.col}"])
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:282]: As noted below, can omit the key func when calling `measure_each` then use `str(q)` here to look up the results.
```suggestion
                    np.mean(results.measurements[str(q)])
```

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 7 above)
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py]: Good catch, thanks!

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 8 above)
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py]: Nice, thanks!

**Comment by eliottrosenberg** [comment]
Thanks for the code reviews, @NoureldinYosri @dstrain115, and @maffoo!

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 10]
```diff
@@ -230,6 +230,63 @@ def single_qubit_randomized_benchmarking(
     return RandomizedBenchMarkResult(num_clifford_range, gnd_probs)
 
 
+def parallel_single_qubit_randomized_benchmarking(
+    sampler: 'cirq.Sampler',
+    qubits: Iterator['cirq.Qid'],
+    use_xy_basis: bool = True,
+    *,
+    num_clifford_range: Sequence[int] = cast(
+        Sequence[int], np.logspace(np.log10(5), 3, 5, dtype=int)
```
Comment (NoureldinYosri) [inline cirq-core/cirq/experiments/qubit_characterizations.py:239]: `np.logspace(np.log10(5), 3, 5, dtype=int)` is an `np.ndarray`. I think there are two options here

1. `num_clifford_range: Sequence[int] = np.logspace(np.log10(5), 3, 5, dtype=int).tolist()`
2. `num_clifford_range: Sequence[int] = tuple(np.logspace(np.log10(5), 3, 5, dtype=int))`

personally I would go with the second option to make sure the default option is immutable.

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 4 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/experiments/qubit_characterizations.py:266]: I think we need better names for these variables. personally I don't know what "cfd" stands for. or at least add a comment explaining what it stands for

**Review by NoureldinYosri — APPROVED** [review]
LGTM. pending review from @dstrain115  and @maffoo

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 11]
```diff
@@ -230,6 +230,63 @@ def single_qubit_randomized_benchmarking(
     return RandomizedBenchMarkResult(num_clifford_range, gnd_probs)
 
 
+def parallel_single_qubit_randomized_benchmarking(
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:233]: OOC, would it be possible to put this in the existing `single_qubit_randomized_benchmarking` module? Seems like it's basically the same thing and there could be opportunities for code reuse. The fact that one is running in parallel on multiple qubits is a minor details (the existing single-qubit code could just become a special case of the parallel code).

Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:227]: Ok, I changed it so that `single_qubit_randomized_benchmarking` is now a wrapper for `parallel_single_qubit_randomized_benchmarking`. What do you think?

Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:227]: I changed it so that `single_qubit_randomized_benchmarking` is a wrapper for a special case of `parallel_single_qubit_randomized_benchmarking` . What do you think?

**Comment by eliottrosenberg** [comment]
It seems that `circuit = circuits.Circuit.zip(*circuits_to_zip)` in `_create_parallel_rb_circuit` is slow, but I'm hoping we can merge this first and then address that in a later PR.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 12]
```diff
@@ -175,9 +175,9 @@ def single_qubit_randomized_benchmarking(
     qubit: 'cirq.Qid',
     use_xy_basis: bool = True,
     *,
-    num_clifford_range: Sequence[int] = range(10, 100, 10),
-    num_circuits: int = 20,
-    repetitions: int = 1000,
+    num_clifford_range: Sequence[int] = tuple(np.logspace(np.log10(5), 3, 5, dtype=int)),
+    num_circuits: int = 10,
+    repetitions: int = 600,
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:180]: Why this change to repetitions? I suspect the runtime for 600 vs 1000 repetitions will be almost indistinguishable since the runtime is likely not dominated by hardware reps.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 13]
```diff
@@ -213,21 +213,73 @@ def single_qubit_randomized_benchmarking(
...
     cliffords = _single_qubit_cliffords()
     c1 = cliffords.c1_in_xy if use_xy_basis else cliffords.c1_in_xz
-    cfd_mats = np.array([_gate_seq_to_mats(gates) for gates in c1])
+    clifford_mats = np.array([_gate_seq_to_mats(gates) for gates in c1])
 
-    gnd_probs = []
-    for num_cfds in num_clifford_range:
-        excited_probs_l = []
+    # create circuits
+    circuits_all = []
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:262]: nit: call this `circuits` and add a type annotation
```suggestion
    circuits: list[`cirq.AbstractCircuit`] = []
```

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 14]
```diff
@@ -213,21 +213,73 @@ def single_qubit_randomized_benchmarking(
...
-            results = sampler.run(circuit, repetitions=repetitions)
-            excited_probs_l.append(np.mean(results.measurements['z']))
-        gnd_probs.append(1.0 - np.mean(excited_probs_l))
+            circuits_all.append(
+                _create_parallel_rb_circuit(qubits, num_cliffords, c1, clifford_mats)
+            )
 
-    return RandomizedBenchMarkResult(num_clifford_range, gnd_probs)
+    # run circuits
+    results_all = sampler.run_batch(circuits_all, repetitions=repetitions)
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:270]: nit: I'd suggest calling this just `results`

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 15]
```diff
@@ -213,21 +213,73 @@ def single_qubit_randomized_benchmarking(
...
+    gnd_probs: dict = {q: [] for q in qubits}
+    idx = 0
+    for num_cliffords in num_clifford_range:
+        excited_probs_l: dict = {q: [] for q in qubits}
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:274]: nit: I'd suggest omitting the `_l` suffix, which is a bit obscure. Also, a full type annotation would be helpful:
```suggestion
        excited_probs: Dict['cirq.Qid', List[float]] = {q: [] for q in qubits}
```

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 16]
```diff
@@ -464,6 +516,18 @@ def _measurement(two_qubit_circuit: circuits.Circuit) -> np.ndarray:
...
+def _create_parallel_rb_circuit(
+    qubits: Iterator['cirq.Qid'], num_cliffords: int, c1: list, clifford_mats: np.ndarray
+) -> 'cirq.Circuit':
+    circuits_to_zip = [
+        _random_single_q_clifford(qubit, num_cliffords, c1, clifford_mats) for qubit in qubits
+    ]
+    circuit = circuits.Circuit.zip(*circuits_to_zip)
+    measure_moment = circuits.Moment(ops.measure_each(*qubits))
+    circuit_with_meas = circuits.Circuit.from_moments(*circuit.moments, measure_moment)
+    return circuit_with_meas
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:528]: nit: can inline measure_each call in the from_moments, and return the final circuit without a local var. Also, iterating over a circuit yields the moments, so don't need to call `.moments` directly:
```suggestion
    return circuits.Circuit.from_moments(*circuit, ops.measure_each(*qubits))
```

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 17]
```diff
@@ -213,21 +213,73 @@ def single_qubit_randomized_benchmarking(
...
+
+def parallel_single_qubit_randomized_benchmarking(
+    sampler: 'cirq.Sampler',
+    qubits: Iterator['cirq.Qid'],
+    use_xy_basis: bool = True,
+    *,
+    num_clifford_range: Sequence[int] = tuple(np.logspace(np.log10(5), 3, 5, dtype=int)),
+    num_circuits: int = 10,
+    repetitions: int = 600,
+) -> dict:
```
Comment (maffoo) [inline cirq-core/cirq/experiments/qubit_characterizations.py:236]: nit: Add full type annotation:
```suggestion
) -> Mapping['cirq.Qid', 'RandomizedBenchmarkResult']:
```

**Review by maffoo — APPROVED** [review]
Minor comments, then LGTM

File: cirq-core/cirq/experiments/qubit_characterizations.py (hunk 13 above)
Comment (eliottrosenberg) [reply cirq-core/cirq/experiments/qubit_characterizations.py:294]: I don't think I should call this `circuits` because earlier in the file, there is `from cirq import circuits`.
